src/ode_solver/reduced_order_ode_solver.cpp

```diff
@@ -0,0 +1,290 @@
+                << std::endl;
+
+    return 0;
+}
+
+template <int dim, typename real, typename MeshType>
+void ReducedOrderODESolver<dim,real,MeshType>::step_in_time (real /*dt*/, const bool /*pseudotime*/)
+{
+    const bool compute_dRdW = true;
+    this->dg->assemble_residual(compute_dRdW);
+
+    this->dg->system_matrix *= -1.0;
+
+    if ((this->ode_param.ode_output) == Parameters::OutputEnum::verbose &&
+        (this->current_iteration%this->ode_param.print_iteration_modulo) == 0 ) {
+        this->pcout << " Evaluating system update... " << std::endl;
+    }
+
+    const Epetra_CrsMatrix epetra_system_matrix = this->dg->system_matrix.trilinos_matrix();
+    const Epetra_CrsMatrix epetra_pod_basis = pod->getPODBasis()->trilinos_matrix();
+    std::shared_ptr<Epetra_CrsMatrix> epetra_test_basis = generate_test_basis(std::make_shared<Epetra_CrsMatrix>(epetra_system_matrix), std::make_shared<Epetra_CrsMatrix>(epetra_pod_basis));
+
+    Epetra_Vector epetra_right_hand_side(Epetra_DataAccess::View, epetra_system_matrix.RowMap(), this->dg->right_hand_side.begin());
+    Epetra_Vector epetra_reduced_rhs(epetra_test_basis->DomainMap());
+    epetra_test_basis->Multiply(true, epetra_right_hand_side, epetra_reduced_rhs);
+    std::shared_ptr<Epetra_CrsMatrix> epetra_reduced_lhs = generate_reduced_lhs(std::make_shared<Epetra_CrsMatrix>(epetra_system_matrix), epetra_test_basis);
+    Epetra_Vector epetra_reduced_solution_update(epetra_reduced_lhs->DomainMap());
+    Epetra_LinearProblem linearProblem(epetra_reduced_lhs.get(), &epetra_reduced_solution_update, &epetra_reduced_rhs);
+
+    Amesos_Lapack Solver(linearProblem);
+    Teuchos::ParameterList List;
+    Solver.SetParameters(List);
+    Solver.SymbolicFactorization();
+    Solver.NumericFactorization();
+    Solver.Solve();
+
+    const dealii::LinearAlgebra::distributed::Vector<double> old_solution(this->dg->solution);
+    double step_length = 1.0;
+    const double step_reduction = 0.5;
+    const int maxline = 10;
+    const double reduction_tolerance_1 = 1.0;
+    const double reduction_tolerance_2 = 2.0;
+
+    double initial_residual;
+    epetra_reduced_rhs.Norm2(&initial_residual);
+    initial_residual /= this->dg->right_hand_side.size();
+    Epetra_Vector epetra_solution(Epetra_DataAccess::View, epetra_pod_basis.RangeMap(), this->dg->solution.begin());
+    Epetra_Vector epetra_solution_update(epetra_pod_basis.RangeMap());
+    epetra_pod_basis.Multiply(false, epetra_reduced_solution_update, epetra_solution_update);
+    epetra_solution.Update(1, epetra_solution_update, 1);
+    this->dg->assemble_residual();
+    epetra_test_basis->Multiply(true, epetra_right_hand_side, epetra_reduced_rhs);
+    double new_residual;
+    epetra_reduced_rhs.Norm2(&new_residual);
+    new_residual /= this->dg->right_hand_side.size();
+
+    // Note that line search is in the same function to avoid having to recompute test basis in a new function or store it as a member variable
```

Comment:
I could, but I would also have to pass a pointer to the POD basis and I would have to modify the class to store epetra_solution_update and epetra_reduced_rhs as member variables. I'd rather avoid this since I used to have epetra vectors as member variables and I think it may have been causing memory leaks.

Reply:
Okay, that's a good reason.